### PR TITLE
Possible feature set mismatch on `fit` and `forecast`

### LINKED ISSUE #560: [BUG] Possible feature set mismatch on `fit` and `forecast`
### 🐛 Bug Report

There is a possible scenario when different sets of features are used on pipeline `fit` and `forecast` stages. The issue arises when the dataset is manipulated/transformed before the pipeline is applied. It is known that using any `IrreversibleTransform`s that add columns before applying the pipeline will lead to an error.
Currently we don't have any checks for such behavior, the model itself would raise an error, which is not always informative enough to resolve this issue.

### Expected behavior

Issue is fixed 
OR 
Sufficient checks and informative errors added

### How To Reproduce

Running the following code will lead to a described error

```python
from etna.datasets import TSDataset
from etna.datasets import generate_ar_df
from etna.transforms import LagTransform
from etna.pipeline import Pipeline
from etna.models import LinearPerSegmentModel


df = generate_ar_df(periods=30)
ts = TSDataset(df=df, freq="D")

transform = LagTransform(in_column="target", lags=[3, 4], out_column="lag")
ts = transform.fit_transform(ts)
print(ts.columns.tolist())

ts = transform.inverse_transform(ts)

pipeline = Pipeline(
    model=LinearPerSegmentModel(),
    transforms=[LagTransform(in_column="target", lags=[5, 6], out_column="lag")],
    horizon=2
)

pipeline.fit(ts)
pipeline.forecast()
```

### Environment

_No response_

### Additional context

It could be related to #440.

### Checklist

- [X] Bug appears at the latest library version

### PR DESCRIPTION
**Pull request by brsnw250**
## Before submitting (must do checklist)

- [x] Did you read the [contribution guide](https://github.com/tinkoff-ai/etna/blob/master/CONTRIBUTING.md)?
- [x] Did you update the docs? We use Numpy format for all the methods and classes.
- [x] Did you write any new necessary tests?
- [x] Did you update the [CHANGELOG](https://github.com/tinkoff-ai/etna/blob/master/CHANGELOG.md)?

## Closing issues

closes #560

### CODE REVIEW

File: etna/datasets/tsdataset.py [hunk 1]
```diff
@@ -509,6 +517,17 @@ def make_future(
...
+
+        additional_columns = (
+            set(self.features)
+            - set(self.target_components_names)
+            - set(self.prediction_intervals_names)
+            - set(future_ts.features)
+            - removed_features
+        )
+        if len(additional_columns) > 0:
+            warnings.warn(f"Some columns were not preserved when building the future dataset: {additional_columns}")
```
Comment (d-a-bunin) [inline etna/datasets/tsdataset.py:529]: There could also be a problem if there are changed columns in `ts`. The `make_future` starts with values before changing, and it could lead to unexpected results. This case seems more difficult to detect.

Comment (d-a-bunin) [reply etna/datasets/tsdataset.py:529]: We could probably write about this in `make_future` docs.

File: tests/test_datasets/test_dataset.py [hunk 2]
```diff
@@ -1095,6 +1095,19 @@ def test_make_future_raise_error_on_diff_endings(ts_diff_endings):
...
 
 
+def test_make_future_additional_columns_warning(ts_diff_endings):
+    transform = LagTransform(lags=[3, 4], in_column="target")
+    transformed_ts = transform.fit_transform(ts=ts_diff_endings)
+    with pytest.warns(UserWarning, match="Some columns were not preserved"):
+        _ = transformed_ts.make_future(2)
+
+
+def test_make_future_with_added_columns(ts_diff_endings):
```
Comment (d-a-bunin) [inline tests/test_datasets/test_dataset.py:1105]: What are we checking here?

File: tests/test_models/test_inference/test_forecast.py [hunk 3]
```diff
@@ -1142,16 +1142,16 @@ class TestForecastMixedInOutSample:
     @staticmethod
     def _test_forecast_mixed_in_out_sample(ts, model, transforms, num_skip_points=50, future_prediction_size=5):
         # fitting
-        df = ts.to_pandas().loc[:, pd.IndexSlice[:, "target"]]
         ts.fit_transform(transforms)
         model.fit(ts)
 
         # forecasting mixed in-sample and out-sample
-        future_ts = ts.make_future(future_steps=future_prediction_size)
```
Comment (d-a-bunin) [inline tests/test_models/test_inference/test_forecast.py:1150]: Why do we need these changes?

Comment (d-a-bunin) [reply tests/test_models/test_inference/test_forecast.py:1150]: The question remains.

File: etna/datasets/tsdataset.py [hunk 4]
```diff
@@ -494,9 +524,30 @@ def make_future(
 
         # Here only df is required, other metadata is not necessary to build the dataset
         ts = TSDataset(df=df, freq=self.freq)
+        removed_features = set(ts.features)
+        expected_to_change = set()
         for transform in transforms:
             tslogger.log(f"Transform {repr(transform)} is applied to dataset")
             transform.transform(ts)
+
+            if hasattr(transform, "in_column"):
```
Comment (d-a-bunin) [inline etna/datasets/tsdataset.py:533]: This logic looks brittle.

File: etna/datasets/tsdataset.py [hunk 5]
```diff
@@ -413,12 +413,42 @@ def _expand_index(df: pd.DataFrame, freq_offset: Optional[pd.DateOffset], future
...
+        if len(same_columns) == 0:
+            return same_columns
+
+        columns_to_use = list(same_columns)
+        df = self._df.loc[:, pd.IndexSlice[:, columns_to_use]]
+        transformed_df = transformed_ts._df.loc[df.index[0] : df.index[-1], pd.IndexSlice[:, columns_to_use]]
+
+        num_df = df.select_dtypes(include=[int, float])
+        transformed_num_df = transformed_df.select_dtypes(include=[int, float])
+        num_mismatch_mask = np.any(~np.isclose(num_df.values, transformed_num_df.values, equal_nan=True), axis=0)
```
Comment (d-a-bunin) [inline etna/datasets/tsdataset.py:428]: What if `num_df` and `transformed_num_df` would have different shape after selecting `int`, `float` dtypes?

File: etna/datasets/tsdataset.py [hunk 6]
```diff
@@ -413,12 +413,42 @@ def _expand_index(df: pd.DataFrame, freq_offset: Optional[pd.DateOffset], future
...
+        num_mismatch = set(num_df.columns[num_mismatch_mask])
+
+        other_df = df.select_dtypes(exclude=[int, float])
+        transformed_other_df = transformed_df.select_dtypes(exclude=[int, float])
+        other_mismatch_mask = np.any(
```
Comment (d-a-bunin) [inline etna/datasets/tsdataset.py:433]: Why not using smth like `other_df.equals(transformed_other_df)`?

Comment (brsnw250) [reply etna/datasets/tsdataset.py:433]: `equals` returns single bool, but we need column mask

File: tests/test_models/test_inference/test_forecast.py (hunk 3 above)
Comment (brsnw250) [reply tests/test_models/test_inference/test_forecast.py:1150]: To include transforms for further forecasting. In other case `make_future` will raise warnings.
This is the similar logic to what was before, just transformation was made earlier.

File: etna/datasets/tsdataset.py (hunk 5 above)
Comment (brsnw250) [reply etna/datasets/tsdataset.py:428]: We work only on intersections here, so this should not be the case. But it may be possible only when a column radically changed its type (e.g. was numeric and became object or categorical). Not sure if it is expected behavior.

File: etna/datasets/tsdataset.py (hunk 4 above)
Comment (d-a-bunin) [reply etna/datasets/tsdataset.py:533]: It looks like a responsibility of a `transform` object for me, not logic of `TSDataset`. But it could be good for now.

**Review by d-a-bunin — APPROVED** [review]
